src/WCSimNuHepMC3Reader.cc

```diff
@@ -0,0 +1,128 @@
+#include "WCSimNuHepMC3Reader.hh"
+
+#include "HepMC3/FourVector.h"
+#include "HepMC3/Print.h"
+#include "HepMC3/ReaderFactory.h"
+
+#include <CLHEP/Random/Random.h>
+#include <G4Types.hh>
+#include <G4ios.hh>
+#include <Randomize.hh>
+#include <cmath>
+
+WCSimNuHepMC3Reader::WCSimNuHepMC3Reader(std::string filename, WCSimDetectorConstruction *myDC)
+    : myDetector(myDC) {
+    G4cout << "NuHepMC3Reader: [INFO] input file called " << filename << G4endl;
+
+    // Get reader object for the input file
+    my_hepmc_reader = HepMC3::deduce_reader(filename);
+
+    G4cout << "NuHepMC3Reader: [INFO] reader object created" << G4endl;
+
+    // Check if the reader is valid
+    if (!my_hepmc_reader) {
+        G4cerr << "NuHepMC3Reader: \033[31m[ERROR]\033[0m Failed to determine input type for HepMC3 file: "
+               << filename << G4endl;
+        exit(1);
+    }
+
+    // Set events parsed to 0
+    events_parsed = 0;
+}
+
+WCSimNuHepMC3Reader::~WCSimNuHepMC3Reader() {}
+
+// Read the next event from the input file
+bool WCSimNuHepMC3Reader::ReadEvent(bool positionFlag) {
+    // Read event
+    my_hepmc_reader->read_event(event);
+
+    // Check if the reading failed
+    if (my_hepmc_reader->failed()) {
+        G4cout << "NuHepMC3Reader: \033[31m[ERROR]\033[0m Failed to read event " << events_parsed << G4endl;
+        // return false;
+    }
+
+    // Set the positon of the vertex to a random position in the detector
+    if (positionFlag) {
+        HepMC3::FourVector position = GenRandomPosition();
+        for (HepMC3::GenParticlePtr part : event.particles()) {
+            if (part->status() == 1) {
+                part->production_vertex()->set_position(position);
+            }
+        }
+    }
+
+    if (!GetParticles()) {
+        return false;
+    };
+
+    events_parsed++;
+    return true;
+}
+
+// Get particles
+bool WCSimNuHepMC3Reader::GetParticles() {
```

Comment:
I don't like the name `GetParticles()`, since the return value isn't `vector<Particle>` or similar

A bad idea is this. Better suggestions are welcomed!
```suggestion
bool WCSimNuHepMC3Reader::ExtractParticlesFromEvent() {
```